tmc-langs-util/src/main/java/fi/helsinki/cs/tmc/langs/util/tarservice/TarCreator.java

```diff
@@ -0,0 +1,99 @@
+     * and creates a tarball.
+     *
+     * @param projectDir     Location of unzipped project dir
+     * @param tmcLangs       Location of tmc-langs-cli.jar
+     * @param tmcrun         Location of tmc-run init script
+     * @param targetLocation Location where the tar archive should be extracted to
+     * @return Returns the location of the folder where the tarball was created
+     * @throws IOException      Error!
+     * @throws ArchiveException Error!
+     */
+
+    public Path createTarFromProject(Path projectDir, Path tmcLangs, Path tmcrun,
+                                     Path targetLocation) throws IOException, ArchiveException {
+        Files.copy(tmcrun, projectDir.resolve(tmcrun.getFileName().toString()));
+        Files.copy(tmcLangs, projectDir.resolve(tmcLangs.getFileName().toString()));
+        createTarBall(projectDir.toFile(), targetLocation);
+        return projectDir.getParent();
+    }
+
+    /**
+     * Creates a tarball from a directory.
+     *
+     * @param project        Project directory file
+     * @param targetLocation Location where the tar archive should be extracted to
+     * @throws IOException      Error!
+     * @throws ArchiveException Error!
+     */
+
+    private void createTarBall(File project, Path targetLocation)
+            throws IOException, ArchiveException {
+        FileOutputStream tarOut;
+        if (targetLocation.getFileName().toString().contains(".tar")) {
+            tarOut = new FileOutputStream(targetLocation.toString());
+        } else {
+            tarOut = new FileOutputStream(targetLocation.toString() + ".tar");
+        }
+        TarArchiveOutputStream aos = new TarArchiveOutputStream(tarOut);
+        aos.setLongFileMode(LONGFILE_POSIX);
+        addFilesToTarBall(project, aos, project.getAbsolutePath().length());
+        aos.finish();
+        aos.close();
+        tarOut.close();
+    }
+
+    /**
+     * Adds all files and folders inside a folder to a tar file.
+     * To not include the path from root to folder, the lengthOfPath integer
+     * should be the same as folder.getPath().length(). To include the full path
+     * insert "" to lengthOfPath
+     *
+     * @param folder       The folder to add
+     * @param tar          TarArchiveOutputStreamer tar
+     * @param lengthOfPath The length of String from root until the start folder.
+     * @throws FileNotFoundException Error!
+     * @throws IOException           Error!
+     */
+
+    private void addFilesToTarBall(File folder, TarArchiveOutputStream tar,
+                                   int lengthOfPath) throws FileNotFoundException, IOException {
+        for (final File file : folder.listFiles()) {
+            if (file.isDirectory()) {
+                addFilesToTarBall(file, tar, lengthOfPath);
+            } else {
+                TarArchiveEntry entry;
+                entry = new TarArchiveEntry(file.toString().substring(lengthOfPath));
```

Comment:
Once you have done the `newDirectoryStream` refactoring, the `file` should be an instance of `Path`. Then you can accomplish relativizing simply by:
```java
basePath.relativize(file);
```
Note that you also need to pass `basePath` from the caller instead of the base path's length.